Add a script that checks static analysis from TravisCI locally.

tammoippen opened the pull request:
This solves #38.

It checks, if the selected executables are working and have the correct version. Then it performs the static code analysis from the TravisCI and gives further information to the user.
It has an incremental mode, which allows to check one file at a time or the user can specify the file he wants to be checked. The CLI arguments follow the implementation from `testsuite/do_tests.sh`.

The documentation can be found in #96 .

heplesser commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,215 @@
+NEST_SRC=.
+CPPCHECK=cppcheck
+VERA=vera++
+VERA_HOME=/usr/local/lib/vera++
```
`VERA_HOME` is not needed, and it would be wrong if `vera++` were installed from MacPorts. One could try to infer it from `which vera++`.

heplesser commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,215 @@
+echo "Tooling:"
+
+# check vera++ is set up appropriately
+$VERA $NEST_SRC/nest/main.cpp >/dev/null 2>&1 || usage 1 "Executable $VERA for vera++ is not working!"
```
On my machine, this command fails because Vera doesn't find the default profile. Although a default profile is in place in `/usr/local/lib/vera++/profiles`. Running with `--profile nest` works.

This test will also fail if one is running outside the NEST source root directory with the `--file` option.

heplesser commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,215 @@
+$CLANG_FORMAT -style=$NEST_SRC/.clang-format $NEST_SRC/nest/main.cpp >/dev/null 2>&1 || usage 1 "Executable $CLANG_FORMAT for clang-format is not working!"
+clang_format_version=`$CLANG_FORMAT --version | sed -${EXTENDED_REGEX_PARAM} 's/^.*([0-9]\.[0-9])\..*/\1/'`
+echo "clang-format version: $clang_format_version"
+if [[ "x$clang_format_version" != "x3.6" ]]; then
```
According to developer space, we require 3.6 *or later*. Why do you require precisely 3.6 here?

tammoippen replied:
There a formatting differences between clang-format 3.7 and 3.6. In TravisCI we use 3.6, so the local test should use the same, to reduce surprises.

heplesser replied:
Could you add a comment to explain this in the script?

heplesser commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,215 @@
+fi
+
+for f in $file_names; do
+  if [ ! -f "$NEST_SRC/$f" ]; then
```
This won't work if one is outside the NEST root directory running with the `--file` option.

heplesser commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,215 @@
+
+    --help               Print program options and exit
+    --incremental        Do analysis one file after another.
+    --file=/path/to/file Perform the static analysis on this file only.
```
Doc should specify that this must be a relative path and that the script must either be run from the nest root directory, or `--nest-src` must also be given. See details on problems that may arise otherwise below.

tammoippen commented:
@heplesser @jougs 

I updated the PR:
• Remove option to set nest src. The script should always be started at nest root.
• Remove VERA_HOME. It was a relict from a former implementation.
• Added documentation in usage ().

heplesser commented:
:+1:

jougs commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,215 @@
+
+      ;;
+    *)
+      echo "$f : not a C/CPP file. Do not do static analysis / formatting checking."
```
"Not executing static analysis / formatting checking." instead of "Do not do ..."

heplesser replied:
Or even better "Skipping ..."

jougs commented:
@abigailm: Could you please check for proper English in the strings in the script and comment?

@tammoippen: After running the tool, I still have several comments before liking this:
* `--help` should also give a one-line explanation of what the tool is good for.
* The tool expects version 3.6 of clang-format. Yet, after installing it from my Ubuntu 15.04 repositories as explained in [NEST Coding Style Guidelines#Tooling](https://nest.github.io/nest-simulator/coding_guidelines_c++#tooling), I get an error that it is not there. Supplying `--clang-format=clang-format-3.6` works, but I wonder if the default should then not be rather `clang-format-3.6` instead of `clang-format`. Or both should be checked if that's depending on the source of the installed package
* Do we really need version 1.69 of `cppcheck` exactly? That's not available in the Ubuntu 15.04 repositories (1.67 is). Either relax the requirement, or remove the line saying that it can be `apt-get` installed in #96, as that apparently does not work.
* Running on an unmodified branch yields an output of "Performing static analysis on:" but no file names (I assume that's because the delta is empty). That output could be "Nothing to check" or something similar instead.

tammoippen commented:
@jougs 
> • --help should also give a one-line explanation of what the tool is good for.

Added with 456b15f.
> • The tool expects version 3.6 of clang-format. Yet, after installing it from my Ubuntu 15.04 repositories ...

Changed with 2b06610. Apparently in both setups (OS X and Ubuntu) there is a clang-format-3.6, so we can very well use that one.
> • Do we really need version 1.69 of cppcheck exactly? That's not available in the ...

I have not done a version by version comparison. The version 1.61, the default version you get with our TravisCI machine,  halts with one source file (see #79). Version 1.69 does not have this error anymore. I changed the documentation with 3b73aa4.
> • Running on an unmodified branch yields an output of "Performing static analysis on:" but no ...

I changed the output message with b490fea.

jougs commented:
@tammoippen: Many thanks for the changes. :+1: from me if @abigailm is OK with the English.

abigailm commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,228 @@
+performed in the same way as in our TravisCI setup. First it looks
+for changed files in the repository in the commit range <git-start>..<git-end>
+(by default this is master..HEAD). Only the changed files are checked.
+You can specify a different commit range or define the file, which
```
You can specify a different commit range, or define the file to be checked, as an argument

abigailm commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,228 @@
+Options:
+
+    --help               Print program options and exit.
+    --incremental        Do analysis one file after another.
```
Do analysis on ...

abigailm commented:
:+1: once the English has been adjusted as suggested by Jochen and me.

heplesser commented:
I added a few more language and comment suggestions; with that, :+1:

heplesser commented on extras/check_code_style.sh:
```diff
@@ -0,0 +1,228 @@
+$VERA --profile nest ./nest/main.cpp >/dev/null 2>&1 ||  bail_out "No profile called nest for vera++ installed. See https://nest.github.io/nest-simulator/coding_guidelines_c++#vera-profile-nest"
+echo "vera++ version: `vera++ --version`"
+
+# check cppcheck 1.69 is installed correctly
```
I think it would be useful to comment in the code why we require 1.69 (and clang 3.6) exactly.

tammoippen commented:
@jougs @heplesser @abigailm 
Ping. I have updated the documentation and help text accordingly.

jougs commented:
I think all concerns were addressed adequately and am thus merging. Thanks for the code and the updates.
